src/err/err_state.rs

```diff
@@ -151,41 +132,75 @@ impl PyErrState {
-        unsafe { ffi::PyErr_Restore(ptype, pvalue, ptraceback) }
+        match self {
+            PyErrState::Lazy(lazy) => {
+                // See note on raise_lazy about possible future efficiency gain
+                raise_lazy(py, lazy);
+                // Safety: raise_lazy will set the raised exception
+                let pvalue = unsafe { Py::from_owned_ptr(py, ffi::PyErr_GetRaisedException()) };
+                PyErrStateNormalized { pvalue }
+            }
+            PyErrState::Normalized(normalized) => normalized,
+        }
     }
 
-    #[cfg(Py_3_12)]
     pub(crate) fn restore(self, py: Python<'_>) {
         match self {
-            PyErrState::Lazy(lazy) => {
-                let PyErrStateLazyFnOutput { ptype, pvalue } = lazy(py);
-                unsafe {
-                    if ffi::PyExceptionClass_Check(ptype.as_ptr()) == 0 {
-                        ffi::PyErr_SetString(
-                            PyTypeError::type_object_raw(py).cast(),
-                            "exceptions must derive from BaseException\0"
-                                .as_ptr()
-                                .cast(),
-                        )
-                    } else {
-                        ffi::PyErr_SetObject(ptype.as_ptr(), pvalue.as_ptr())
-                    }
-                }
+            PyErrState::Lazy(lazy) => raise_lazy(py, lazy),
+            #[cfg(not(Py_3_12))]
+            PyErrState::FfiTuple {
+                ptype,
+                pvalue,
+                ptraceback,
+            } => {
+                let ptype = ptype.into_ptr();
+                let pvalue = pvalue.map_or(std::ptr::null_mut(), Py::into_ptr);
+                let ptraceback = ptraceback.map_or(std::ptr::null_mut(), Py::into_ptr);
+                unsafe { ffi::PyErr_Restore(ptype, pvalue, ptraceback) }
             }
-            PyErrState::Normalized(PyErrStateNormalized { pvalue }) => unsafe {
-                ffi::PyErr_SetRaisedException(pvalue.into_ptr())
+            PyErrState::Normalized(PyErrStateNormalized {
+                #[cfg(not(Py_3_12))]
+                ptype,
+                pvalue,
+                #[cfg(not(Py_3_12))]
+                ptraceback,
+            }) => unsafe {
+                #[cfg(not(Py_3_12))]
+                {
+                    ffi::PyErr_Restore(
+                        ptype.into_ptr(),
+                        pvalue.into_ptr(),
+                        ptraceback.map_or(std::ptr::null_mut(), Py::into_ptr),
+                    )
+                }
+                #[cfg(Py_3_12)]
+                {
+                    ffi::PyErr_SetRaisedException(pvalue.into_ptr())
+                }
             },
         }
     }
 }
+
+/// Raises a "lazy" exception state into the Python interpreter.
+///
+/// In principle this could be split in two; first a function to create an exception
+/// in a normalized state, and then a call to `PyErr_SetRaisedException` to raise it.
```

Comment:
See also this comment; I think in principle we could implement an equivalent of `PyErr_SetObject` in Rust, which would have the upside of allowing us to then refactor and adjust behaviour later if we so desire. I wanted to start by just calling `PyErr_SetObject` as it avoids moving a lot of complexity to Rust.